text/projects/Maciej_Nems/Simple_String_Matching_k_Mistakes.tex

```diff
@@ -0,0 +1,126 @@
+\par
+Algorytm dla każdego znaku w tekście najpierw go dorzuca do kolejki i dekrementuje licznik. Jeśli $C[znak] < 0$, to znaczy, że wykorzystaliśmy już wszystkie wystąpienia tego znaku w $W$, więc mamy błąd (inkrementujemy $K$). Teraz tak długo, aż błędów w kolejce jest ponad $k$, usuwamy znaki z kolejki i odpowiednio dekrementujemy $K$. Po pętli, jeśli rozmiar $Q$ równa się $m$, to zebraliśmy tyle znaków, co długość wzorca, mamy co najwyżej $K$ błędów, czyli wiemy, że to jest permutacja wzorca z co najwyżej $K$ błędami i możemy ją zgłosić jako match. Dodatkowo, aby zawsze na początku pętli rozmiar $Q$ wynosił co najwyżej $m-1$, to usuwamy pierwszy element.
+\begin{algorithm}[H]
+\caption{Permutation Matching}
+\begin{algorithmic} 
+\INPUT{$T$, $W$, $n$, $m$, $k$}
+\IF{$n < m$}
+    \STATE{end with no matches}
+\ENDIF
+\STATE{\textbf{Preprocessing:} For each $a \in A_W$ set $C[a]$ to count of $a$ in $W$; Substitute all $t \in T$, $t \notin A_T$ with special symbol $\# \notin A_P$ to get $\tilde{T}$; Set $C[\#] = 0$}
+\STATE{$Z:=0$}
+\FOR{$j=1$ \TO $n$}
+    \STATE{$Q.push(\tilde{T}[j])$}
+    \STATE{$C[\tilde{T}[j]] = C[\tilde{T}[j]] - 1$}
+    \IF{$C[\tilde{T}j]] < 0$}
+        \STATE{$Z = Z + 1$}        
+    \ENDIF\\
+    \STATE{// Pop from $Q$, until there are at most $k$
+    mismatches}
+    \WHILE{$Z > k$}
+        \STATE{$x = Q.pop()$}
+        \IF{$C[x] < 0$}
+            \STATE{$Z = Z - 1$}
+        \ENDIF
+        \STATE{$C[X] = C[X] + 1$}
+    \ENDWHILE
+    \STATE{// If $Q$ has length $m$ report occurance, as there are at most $k$ mismatches}
+    \STATE{// also prepare for next iteration by poping from $Q$}
+    \IF{$Q.size() == m$}
+        \STATE{Report occurence in position $j - m + 1$}
+        \STATE{$x = Q.pop()$}
+        \IF{$C[x] < 0$}
+            \STATE{$Z = Z - 1$}
+        \ENDIF
+        \STATE{$C[X] = C[X] + 1$}
+    \ENDIF
+\ENDFOR
+\end{algorithmic}
+\end{algorithm}
+\subsubsection*{Poprawność}
+Dość łatwo zauważyć, że algorytm ten jest poprawny. W każdej iteracji na koniec sprawdzamy warunki konieczne, by pozycja $j-m+1$ była dopasowaniem. Tzn sprawdzamy liczbę zebranych znaków w kolejce i liczbę błędów. Przesuwamy się w każdym wykonaniu pętli o $1$, więc sprawdzimy wszystkie możliwe pozycje.
+\subsubsection*{Złożoność}
+Złożoność takiego dopasowania to $O(n\log |A_P|$). Musimy stworzyć słownik, który zawiera $A_P$ oraz $\#$ w czasie $O(|A_P|\log |A_P|)$. Do kolejki $Q$ dodamy co najwyżej $n$ razy więc i odejmiemy co najwyżej $n$ razy. Oznacza to, że sumarycznie główna pętla zajmie $O(n\log |A_P|)$ czasu, ponieważ $n$ razy zrobimy operacje liniowe typu dodaj do kolejki, odejmij z kolejki, inkrementuj, dekrementuj, oraz $n$ razy zrobimy operacje logarytmiczne dostępu do słownika $C$.
```

Comment:
Well, in practice yes. But isn't using `O(1)` for hashing table in algorithm time complexity proof a little bit of a stretch? Technically it can have `O(|A_P|)` or `O(log |A_P|)` in worst case scenario.

Reply:
There are hashing tables with $O(1)$ worst-case lookup e.g. https://en.wikipedia.org/wiki/Cuckoo_hashing (in fact any perfect hashing should be sufficient) - the cost is insertions are only expected constant time, but here it averages out since we build $C$ once.